Source/Particles/ParticleBoundaries.cpp

```diff
@@ -8,6 +8,61 @@
 #include "ParticleBoundaries.H"
 
 #include "Utils/Parser/ParserUtils.H"
+#include "Utils/TextMsg.H"
+
+namespace warpx::particles
+{
+    std::pair<
+        amrex::Array<ParticleBoundaryType, AMREX_SPACEDIM>,
+        amrex::Array<ParticleBoundaryType, AMREX_SPACEDIM>
+    >
+    parse_particle_boundaries (
+        const amrex::Array<FieldBoundaryType, AMREX_SPACEDIM>& field_boundary_lo,
+        const amrex::Array<FieldBoundaryType, AMREX_SPACEDIM>& field_boundary_hi)
+    {
+        auto particle_boundary_lo =
+            amrex::Array<ParticleBoundaryType, AMREX_SPACEDIM>{
+                AMREX_D_DECL(
+                    ParticleBoundaryType::Default,
+                    ParticleBoundaryType::Default,
+                    ParticleBoundaryType::Default)};
+        auto particle_boundary_hi = particle_boundary_lo;
+
+        const auto pp_boundary = amrex::ParmParse{"boundary"};
+
+        const bool particle_boundary_specified =
+            pp_boundary.contains("particle_lo") ||
+            pp_boundary.contains("particle_hi");
+
+            for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+                pp_boundary.query_enum_sloppy("particle_lo", particle_boundary_lo[idim], "-_", idim);
+                pp_boundary.query_enum_sloppy("particle_hi", particle_boundary_hi[idim], "-_", idim);
+
+                if (field_boundary_lo[idim] == FieldBoundaryType::Periodic ||
+                    field_boundary_hi[idim] == FieldBoundaryType::Periodic ||
+                    particle_boundary_lo[idim] == ParticleBoundaryType::Periodic ||
+                    particle_boundary_hi[idim] == ParticleBoundaryType::Periodic ) {
+                    // to ensure both lo and hi are set to periodic consistently for both field and particles.
+                    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+                        (field_boundary_lo[idim]  == FieldBoundaryType::Periodic) &&
+                        (field_boundary_hi[idim]  == FieldBoundaryType::Periodic),
+                        "field boundary must be consistenly periodic in both lo and hi");
+                    if (particle_boundary_specified) {
+                        WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+                            (particle_boundary_lo[idim] == ParticleBoundaryType::Periodic) &&
+                            (particle_boundary_hi[idim] == ParticleBoundaryType::Periodic),
+                            "field and particle boundary must be periodic in both lo and hi");
+                    } else {
+                        // set particle boundary to periodic
+                        particle_boundary_lo[idim] = ParticleBoundaryType::Periodic;
+                        particle_boundary_hi[idim] = ParticleBoundaryType::Periodic;
+                    }
+                }
```

Comment:
I wonder if it would be clearer to have `parse_particle_boundaries` take no arguments and skip the check, and define a separate function, say, `check_periodicity` or something along these lines, that takes both the field and particle boundary tuples, and performs the periodicity check.

This separate function would be then called both in WarpX.cpp and Initialization/WarpXAMReXInit.cpp (from within `set_periodicity_according_to_boundary_types`).

This would seem clearer to me, from a functional point of view, but it's just a suggestion. Let me know what you think or if you see any drawbacks.

Reply:
@lucafedeli88

I think we are close to finalizing this PR, as soon as we resolve the last comment here. Let me know what you think!

Reply:
Thanks for your feedback @EZoni , and sorry for my late answer. 
I would like to propose a slightly different solution, since field boundary conditions are required to select the default value of the boundary conditions for particles (i.e., they are not just used for consistency checks): 

- consistency checks are moved to helper functions hidden in the implementation file, for clarity
- a new helper function generates a "periodicity array" for field boundary conditions (true = periodic, false = otherwise )
- this bool array is now the argument of the function that reads particles' boundary conditions. 

What do you thinks about this solution ?

Reply:
Oh I just noticed that you already pushed more commits to refactor things! Will review the changes again.